weave/workers/weave_scorer.py

```diff
@@ -0,0 +1,458 @@
+        filter=tsi.ObjectVersionFilter(
+            is_op=False,
+            base_object_classes=[Monitor.__name__],
+            latest_only=True,
+        ),
+    )
+
+    server = get_trace_server()
+
+    monitor_objects = server.objs_query(obj_query)
+
+    active_monitors = [
+        ActiveMonitor(
+            monitor=Monitor(
+                name=obj.val["name"],
+                description=obj.val["description"],
+                sampling_rate=obj.val["sampling_rate"],
+                scorers=resolve_scorer_refs(obj.val["scorers"], project_id),
+                op_names=obj.val["op_names"],
+                query=obj.val["query"],
+                active=obj.val["active"],
+            ),
+            internal_ref=InternalObjectRef(
+                project_id=project_id, name=obj.val["name"], version=obj.digest
+            ),
+            wb_user_id=obj.wb_user_id,
+        )
+        for obj in monitor_objects.objs
+        if obj.val["active"]
+    ]
+
+    return active_monitors
+
+
+def resolve_scorer_refs(scorer_ref_uris: list[str], project_id: str) -> list[Scorer]:
+    """Resolves scorer references to Scorer objects."""
+    server = get_trace_server()
+
+    scorer_refs = [
+        parse_internal_uri(scorer_ref_uri) for scorer_ref_uri in scorer_ref_uris
+    ]
+
+    scorer_dicts = server.refs_read_batch(
+        tsi.RefsReadBatchReq(refs=scorer_ref_uris)
+    ).vals
+
+    scorers = [
+        maybe_objectify(from_json(scorer_dict, project_id, server))
+        for scorer_dict in scorer_dicts
+    ]
+
+    for scorer, scorer_ref in zip(scorers, scorer_refs):
+        scorer.__dict__["internal_ref"] = scorer_ref
+
+    return scorers  # type: ignore
+
+
+class CallNotWrittenError(Exception):
+    pass
+
+
+# We do the following retry logic because ClickHouse does not guarantee that the call will be available immediately.
+@retry(
+    retry=retry_if_exception_type(CallNotWrittenError),
+    wait=wait_fixed(1),
+    stop=stop_after_attempt(5),
+    before=before_log(logger, logging.INFO),
+)
+def get_filtered_call(
+    op_names: list[str],
+    query: Optional[tsi.Query],
+    ended_call: tsi.EndedCallSchemaForInsert,
+) -> Optional[Call]:
+    """Looks up the call based on a monitor's call filter."""
+    server = get_trace_server()
+
+    # We do this two-step querying to circumvent the absence of write->read consistency in ClickHouse.
+    # We want to differentiate between a call not written yet and a call existing but not matching the filter.
+    # - first we count the calls that match the call id
+    # - if that returns 0, we raise an exception in order to trigger a retry
+    # - if that returns > 0, we know the call exists, so we can proceed with the filter
+
+    count_req = tsi.CallsQueryStatsReq(
+        project_id=ended_call.project_id,
+        filter=tsi.CallsFilter(call_ids=[ended_call.id]),
+    )
+
+    count = server.calls_query_stats(count_req).count
+
+    if count == 0:
```

Comment:
assuming that this will return 0 when the start event is not yet written, then this will effectively serve as a "wait" for the start event

Reply:
this wait however locks up the worker thread until complete and possibly rejects before the start comes through